plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageService.java

```diff
@@ -185,6 +194,56 @@ public void deleteBlob(String account, String container, String blob) throws URI
         });
     }
 
+    void deleteBlobDirectory(String account, String container, String path, Executor executor)
+            throws URISyntaxException, StorageException, IOException {
+        final Tuple<CloudBlobClient, Supplier<OperationContext>> client = client(account);
+        final CloudBlobContainer blobContainer = client.v1().getContainerReference(container);
+        final Collection<Exception> exceptions = Collections.synchronizedList(new ArrayList<>());
+        final AtomicLong outstanding = new AtomicLong(0);
+        final SetOnce<PlainActionFuture<Void>> resReference = new SetOnce<>();
+        SocketAccess.doPrivilegedVoidException(() -> {
+            for (final ListBlobItem blobItem : blobContainer.listBlobs(path, true)) {
+                // uri.getPath is of the form /container/keyPath.* and we want to strip off the /container/
+                // this requires 1 + container.length() + 1, with each 1 corresponding to one of the /
+                final String blobPath = blobItem.getUri().getPath().substring(1 + container.length() + 1);
+                outstanding.incrementAndGet();
+                executor.execute(new AbstractRunnable() {
+                    @Override
+                    protected void doRun() throws Exception {
+                        deleteBlob(account, container, blobPath);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        exceptions.add(e);
+                    }
+
+                    @Override
+                    public void onAfter() {
+                        if (outstanding.decrementAndGet() == 0) {
+                            ActionListener<Void> resListener = resReference.get();
+                            if (resListener != null) {
+                                resListener.onResponse(null);
+                            }
+                        }
+                    }
+                });
+            }
+        });
+        if (outstanding.get() > 0) {
+            final PlainActionFuture<Void> result = PlainActionFuture.newFuture();
+            resReference.set(result);
```

Comment:
I figured it's safe here since we do all the incrementing in the same thread when executing the loop, so we can't accidentally run into `0` so I figured I don't need to increment before the loop?

Reply:
I find the following easier to understand:

```
diff --git a/plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageService.java b/plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageService.java
index 850fb9d3fad..187ccd760c3 100644
--- a/plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageService.java
+++ b/plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageService.java
@@ -199,8 +199,8 @@ public class AzureStorageService {
         final Tuple<CloudBlobClient, Supplier<OperationContext>> client = client(account);
         final CloudBlobContainer blobContainer = client.v1().getContainerReference(container);
         final Collection<Exception> exceptions = Collections.synchronizedList(new ArrayList<>());
-        final AtomicLong outstanding = new AtomicLong(0);
-        final SetOnce<PlainActionFuture<Void>> resReference = new SetOnce<>();
+        final AtomicLong outstanding = new AtomicLong(1);
+        final PlainActionFuture<Void> result = PlainActionFuture.newFuture();
         SocketAccess.doPrivilegedVoidException(() -> {
             for (final ListBlobItem blobItem : blobContainer.listBlobs(path, true)) {
                 // uri.getPath is of the form /container/keyPath.* and we want to strip off the /container/
@@ -221,22 +221,16 @@ public class AzureStorageService {
                     @Override
                     public void onAfter() {
                         if (outstanding.decrementAndGet() == 0) {
-                            ActionListener<Void> resListener = resReference.get();
-                            if (resListener != null) {
-                                resListener.onResponse(null);
-                            }
+                            result.onResponse(null);
                         }
                     }
                 });
             }
         });
-        if (outstanding.get() > 0) {
-            final PlainActionFuture<Void> result = PlainActionFuture.newFuture();
-            resReference.set(result);
-            if (outstanding.get() > 0) {
-                result.actionGet();
-            }
+        if (outstanding.decrementAndGet() == 0) {
+            result.onResponse(null);
         }
+        result.actionGet();
         if (exceptions.isEmpty() == false) {
             final IOException ex = new IOException("Deleting directory [" + path + "] failed");
             exceptions.forEach(ex::addSuppressed);
```

Reply:
Fair point, applied your version :)